stream_alert/threat_intel_downloader/threat_stream.py

```diff
@@ -0,0 +1,260 @@
+
+class ThreatStream(object):
+    """Class to retrieve IOCs from ThreatStream.com and store them in DynamoDB"""
+    _API_URL = 'https://api.threatstream.com'
+    _API_RESOURCE = 'intelligence'
+    _IOC_STATUS = 'active'
+    # max IOC objects received from one API call, default is 0 (equal to 1000)
+    _API_MAX_LIMIT = 1000
+    _API_MAX_INDEX = 1000000
+    _PARAMETER_NAME = 'threat_intel_downloader_api_creds'
+    _MAX_RETRY = 3
+
+    def __init__(self, config):
+        self.ioc_types = config['ioc_types']
+        self.ioc_sources = config['ioc_filters']
+        self.threshold = self._API_MAX_INDEX - self._API_MAX_LIMIT
+        self.region = config['region']
+        self.ioc_keys = config['ioc_keys']
+        self.api_user = None
+        self.api_key = None
+        self._get_api_creds()
+        self.table_name = config['function_name']
+
+    def _get_api_creds(self):
+        """Retrieve ThreatStream API credentials from Parameter Store"""
+        try:
+            ssm = boto3.client('ssm', self.region)
+            response = ssm.get_parameters(
+                Names=[self._PARAMETER_NAME], WithDecryption=True
+            )
+            for cred in response['Parameters']:
+                if cred['Name'] == self._PARAMETER_NAME:
+                    try:
+                        decoded_creds = json.loads(cred['Value'])
+                        self.api_user = decoded_creds['api_user']
+                        self.api_key = decoded_creds['api_key']
+                    except ValueError:
+                        LOGGER.debug('Can not load value for parameter with '
+                                     'name \'%s\'. The value is not valid json: '
+                                     '\'%s\'', cred['Name'], cred['Value'])
+                        raise ThreatStreamCredsError
+        except ClientError as err:
+            LOGGER.debug('SSM client error: %s', err)
+            raise
+
+        if not (self.api_user and self.api_key):
+            LOGGER.debug('API Creds Error')
+            raise ThreatStreamCredsError
+
+
+    def _connect(self, next_url):
+        """Send API call to ThreatStream with next token and return parsed IOCs
+
+        The API call has retry logic up to 3 times.
+        Args:
+            next_url (str): url of next token to retrieve more objects from
+                ThreatStream
+
+        Returns:
+            (tuple): (list, str, bool)
+                - First object is a list of intelligence.
+                - Second object is a string of next token to retrieve more IOCs.
+                - Third object is bool to indicated if retrieve more IOCs from
+                    threat feed.
+                    Return False if next token is empty or threshold of number
+                    of IOCs is reached.
+        """
+        retry, retry_num, continue_invoke = True, self._MAX_RETRY, False
+        intelligence = list()
+        while retry:
```

Comment:
I think a good amount of this complex logic can be handled using the `backoff` library via a few stacked decorators. Maybe @jacknagz would be willing to offer some pointers here?

Reply:
Thanks! I looked `backoff` library last night, it is very useful. I will include it in my next commit.